File: homework2/quickSort.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> nums;
unsigned long long calledTimes;

void saveNums();
void quickSort(int leftLimit, int rightLimit);
int pivotPartition(const int leftMostIndex, const int rightMostIndex);

int main()
{
    freopen("in.txt", "r", stdin);
    saveNums();
    quickSort(0, nums.size());
    cout << calledTimes << endl;
    return 0;
}

void saveNums()
{
    int num;
    while (cin >> num)
        nums.push_back(num);
}

void quickSort(int leftLimit, int rightLimit)
{
    if ((rightLimit - leftLimit) <= 1)
        return;

    calledTimes += (rightLimit - leftLimit) - 1;
    int pivotPosition = pivotPartition(leftLimit, rightLimit); //  if pivot is not in the first, i could change it

    quickSort(leftLimit, pivotPosition - 1);
    quickSort(pivotPosition + 1, rightLimit);
    // recurse first
    // recurse second
}

// que pasa si es la ultima pos del array
int pivotPartition(const int leftMostIndex, const int rightMostIndex)
{
    const int pivot = nums[leftMostIndex];
    int pivotEndingLocation = leftMostIndex; // to make sure that is rearenged
    for (int j = leftMostIndex + 1; j < rightMostIndex; ++j)
    {
        // im not considering an equals case
        if (nums[j] < pivot)
        {
            pivotEndingLocation += 1;
            int temp = nums[j];
            nums[j] = nums[pivotEndingLocation];
            nums[pivotEndingLocation] = temp;
        }
    }
    int temp = nums[pivotEndingLocation];
    nums[pivotEndingLocation] = pivot;
    nums[leftMostIndex] = temp;
    return pivotEndingLocation;
}

File: homework2/quickSort.py

calls = 0
i = 1


def medianOfTheTree(nums, leftLimit, rightLimit):
    size = (rightLimit + leftLimit)
    print('left', 'rigth', 'size', leftLimit, rightLimit, size)
    midIndex = size // 2
    if size % 2 == 0:
        midIndex -= 1

    firstNum = nums[leftLimit]
    midNum = nums[midIndex]
    secondNum = nums[rightLimit-1]

    pivot = sorted([firstNum, midNum, secondNum])[1]

    index = -1

    if firstNum == pivot:
        index = leftLimit
    elif midNum == pivot:
        index = midIndex
    else:
        index = rightLimit-1

    nums[leftLimit], nums[index] = nums[index], nums[leftLimit]


def pivotPartition(nums, leftLimit, rightLimit):
    pivot = nums[leftLimit]
    pivotEndingLocation = leftLimit
    for i in range(leftLimit + 1, rightLimit):
        if nums[i] < pivot:
            pivotEndingLocation += 1
            nums[i], nums[pivotEndingLocation] = nums[pivotEndingLocation], nums[i]
    nums[pivotEndingLocation], nums[leftLimit] = pivot, nums[pivotEndingLocation]
    return pivotEndingLocation


def quickSort(nums, leftLimit, rightLimit):
    if rightLimit - leftLimit <= 1:
        return

    global calls
    calls += (rightLimit - leftLimit) - 1

    # nums[rightLimit-1],nums[leftLimit]  = nums[leftLimit],nums[rightLimit-1]
    medianOfTheTree(nums, leftLimit, rightLimit)     # for the second problem

    pivotPosition = pivotPartition(nums, leftLimit, rightLimit)
    quickSort(nums, leftLimit, pivotPosition)
    quickSort(nums, pivotPosition + 1, rightLimit)


def main():
    file = open('inVid.txt', 'r')
    nums = []
    for line in file.readlines():
        nums.append(int(line))
    quickSort(nums, 0, len(nums))

    global calls
    print(calls)

    file.close()


if __name__ == '__main__':
    main()
